From 9dc9c2a0b714b01e6af0714b6fcace4b5b4bb0de Mon Sep 17 00:00:00 2001
From: Abderrahmane Benbachir <abderrahmane.benbachir@polymtl.ca>
Date: Sun, 23 Jul 2017 11:29:16 -0400
Subject: [PATCH] KVM: x86: Fix vmcall exit multiplication

This optimization make a single vmcall exit fast and reduce frequency of exits.

Signed-off-by: Abderrahmane Benbachir <abderrahmane.benbachir@polymtl.ca>
---
 arch/x86/kvm/vmx.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index a236decb81e4..b81141a315c9 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8204,7 +8204,19 @@ static bool nested_vmx_exit_handled(struct kvm_vcpu *vcpu)
 		return nested_cpu_has(vmcs12, CPU_BASED_RDPMC_EXITING);
 	case EXIT_REASON_RDTSC: case EXIT_REASON_RDTSCP:
 		return nested_cpu_has(vmcs12, CPU_BASED_RDTSC_EXITING);
-	case EXIT_REASON_VMCALL: case EXIT_REASON_VMCLEAR:
+	case EXIT_REASON_VMCALL:
+		/*
+		 * Exit multiplication : This fix prevent VMCALL from being
+		 * forwarded to L1, because a single L2 exit can cause many
+		 * L1 exits!, this optimization make a single vmcall exit
+		 * fast and reduce frequency of exits.
+		 * Only vmcall with specific hypercall number range will
+		 * benifit from this optimization.
+		*/	
+		if (kvm_register_read(vcpu, VCPU_REGS_RAX) < 0x10000)
+			return false;
+		return true;
+	case EXIT_REASON_VMCLEAR:
 	case EXIT_REASON_VMLAUNCH: case EXIT_REASON_VMPTRLD:
 	case EXIT_REASON_VMPTRST: case EXIT_REASON_VMREAD:
 	case EXIT_REASON_VMRESUME: case EXIT_REASON_VMWRITE:
-- 
2.11.0

